appsrc/dconvert/imtn/imtntodc.man
.TH IMTNTODC 1 "06 March 2014" "DICOM PS3" "Imatron to DICOM"
.SH NAME
imtntodc \- ACR/NEMA DICOM PS3 ... Imatron to DICOM
.SH SYNOPSIS
.HP 10
.B imtntodc
.so man1/gen.so
[
.B \-v|verbose
]
.I inputfilename
.I outputfilename
.SH DESCRIPTION
.LP
.B imtntodc
reads the named Imatron input file and converts it to a dicom output file.
.LP
.SH OPTIONS
The verbose output goes to standard error.
.PP
The basic switches are described in dcintro(1).
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
\ 
.RE
% imtntodc \-v test test.dc3
.RE
\ 
.RE
Converting ...
.RE
******** As read ... ********
.RE
(0x0008,0x0008) CS Image Type ...
.RE
******** As written ... ********
.RE
(0x0002,0x0001) OB File Meta Information Version ...
.RE
 ...
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR dcintro(1) ,
.BR imtndump(1)
.SH AUTHOR
Copyright (c) 1993-2020, David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS


CHANGES
210306:	libsrc/standard/condn.tpl,iodcomp,module,strval/vl.tpl: Add Dermoscopic Photography Image (Sup 221)

210306:	libsrc/standard/binval.tpl,iodcomp,module,strval/vl.tpl: Add Visible Light Photography Attributes corresponding to EXIF 2.31 and TIFF/EP (CP 1736)

210206:	libsrc/standard/condn.tpl: Include SCT codes when determining Laterality Required, using 2021a as refernce, including some updated codes (000555)

210121:	libsrc/standard/elmdict/dicom3.tpl: Add new data elements for XA Protocol (Sup 212)

210119:	libsrc/standard/elmdict/dicom3.tpl: Add new data elements (CPACK 109)

210104:	libsrc/standard/elmdict/other.tpl: Update PixelMed private WSI bulk annotation data elements with Annotation Group UID (sup222_08_WSIAnnotations)

210102:	libsrc/standard/elmdict/dicom3.tpl,module/rt.tpl: Source Pixel Plane Characteristics for Contours (CP 2006); Extend RT Structure Set ROI Color (CP 1395)

201208:	appsrc/dcfile/dciodvfy.cc: Check sanity of tile dimensions and number of frames in whole slide images (000553)

201208: libsrc/standard/elmdict/other.tpl: Add more observed 3DHISTECH private data elements

201208: libsrc/standard/condn.tpl: Do not report zero length ViewModifierCodeSequence error for non Mammo DX when ViewModifierCodeSequence not present (regression caused by 000549) (000552)

201208: libsrc/standard/condn.tpl: Do not complain about presence of ReferencedSeriesSequence or StudiesContainingOtherReferencedInstancesSequence (Common Instance Reference Module) if SegmentSurfaceSourceInstanceSequence is present (000551)

201208: libsrc/src/dctool/condn.cc,standard/condn.tpl,support/condn.awk: Do not complain about absence of ReferencedSeriesSequence and StudiesContainingOtherReferencedInstancesSequence (Common Instance Reference Module) when valid empty (no item) ReferencedImageSequence (000550)

201205: libsrc/standard/condn.tpl,module/dx.tpl: Do not report error on zero length ViewModifierCodeSequence if Mammo, since Type 2 specializes DX Positioning Module Type 3 (regression caused by change on 2013/06/26) (000549)

201117:	libsrc/standard/elmdict/dicom3.tpl: Add OV, SV and UV Selector Attributes (CP 2042)

201117:	libsrc/standard/elmdict/dicom3.tpl: SR Table data elements (CP 2041)

201115:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Dermoscopy data elements (Sup 221)

200920:	standard/module/base.tpl,module/base.tpl: Make Frame Anatomy in Parametric Map Optional (CP 2004)

200919:	libsrc/standard/condn.tpl,elmdict/dicom3.tpl,module/base.tpl,us.tpl: Add ultrasound Probe ID in Transducer Identification Sequence (CP 1951); depended on adding Device Identification Sequence Macro (from Sup 147)

200915:	libsrc/standard/elmdict/other.tpl: Update PixelMed private WSI bulk annotation data elements (supnnn_05_WSIAnnotations)

200829:	libsrc/standard/condn.tpl,iodcomp,module/base.tpl: Check CommonInstanceReference Module presence conditional on references being present (000547) and attributes only present when references are present (000548)

200822:	libsrc/standard/elmdict/dicom3.tpl: Render MakeNote as string even though OB

200806:	standard/condn.tpl,module/base.tpl,mr.tpl,vl.tpl: Do not complain about YBR_RCT for .90 Transfer Syntax (000546)

200726:	libsrc/standard/strval/vl.tpl: Correct NUMERIC value type in Acquisition Context (000545)

200724:	libsrc/standard/module/mr.tpl: Fix incorrect condition for Real World Value Mapping Macro in Shared Functional Group Sequence for Legacy Converted MR (000544)

200719:	libsrc/standard/condn.tpl,elmdict/dicom3.tpl,diconde.tpl: Correct VRs, VMs and keywords to match 2020c (Peros)

200716:	libsrc/standard/elmdict/siemens.tpl: Correct duplicate and incorrect private data elements (Vrinda, Conformance Statement)

200716:	libsrc/standard/elmdict/gems.tpl: Correct duplicate and incorrect private data elements (Vrinda, Conformance Statement)

200716:	libsrc/standard/elmdict/agfa.tpl: Correct duplicate and incorrect private data elements (Vrinda, Conformance Statement)

200716:	libsrc/standard/condn.tpl: Correct KVP CT Image Module condition when multi-energy (000543)

200716:	libsrc/standard/module/ct.tpl: Use ReferencedXRaySourceIndex not ReferencedPathIndex for multi-energy CT Exposure Sequence (000542)

200716:	libsrc/standard/elmdict/dicom3.tpl: Add data elements and SOP Class UIDs for latest release (includng catching up on 2nd Gen RT UIDs) (2020c)

200622:	libsrc/include,src/dctool/attrtypo.h,.cc: Allow insertion of strings in OB using addValues(), e.g., for dccp (000541)

200620:	libsrc/src/dctool/dcopto.cc: Allow insertion of empty sequence during dccp (000540)

200610:	libsrc/standard/elmdict/philips.tpl,siemens.tpl: Update observed private data elements, including OB rendered as string for some

200610:	libsrc/include/dctool/ie.h,src/dctool/ie.cc,support/module.awk: Fix dcentvfy SeriesInstanceUID broken by addition of ReferencedInstancesAndAccessMacro for Patient Photo by overwriting IE for attribute set during module parsing when depth in the hierarchy is lower (000539)

200608:	libsrc/standard/condn.tpl,elmdict/dicom3.tpl,module,strval/base.tpl: Update Encapsulated Document Module (CP 1811 and Sup 208, including new data elements)

200605:	libsrc/standard/elmdict/other.tpl: Add Merge Unity PACS (DR Systems) data elements as described in DCS v11.2 and encountered

200605:	libsrc/standard/elmdict/other.tpl: Add Philips/Stentor iSite data elements as encountered

200604:	libsrc/standard/elmdict/other.tpl: Add Samsung CR private data elements as encountered, as UT

200604:	libsrc/standard/elmdict/gems.tpl: More private GEMS_PARM_01 group 0043 GE MR data elements from DV26, PX26.0, SV25.1, PX26.1 DCS

200526:	libsrc/standard/elmdict/elscint.tpl: Correct VR of (5001,0070) from SQ to US and other related elements based on observed GSPS - may be conflict with ELSCINT1 creator [fragile]

200520:	libsrc/standard/elmdict/toshiba.tpl: Add more private data elements from CT Conformance Statements


appsrc/dconvert/imtn/imtntodc.cc
static const char *CopyrightIdentifier(void) { return "@(#)imtntodc.cc Copyright (c) 1993-2020, David A. Clunie DBA PixelMed Publishing. All rights reserved."; }
#include "attrmxls.h"
#include "transynu.h"
#include "mesgtext.h"
#include "bnopt.h"
#include "dcopt.h"

#include "imtn.h"

int
main(int argc,char **argv)
{
	GetNamedOptions 	options(argc,argv);
	BinaryInputOptions 	binary_input_options(options);
	DicomOutputOptions 	dicom_output_options(options);

	bool verbose=options.get("verbose") || options.get("v");

	int imagenumber;
	if (!options.get("image",imagenumber)) imagenumber=0;

	binary_input_options.done();
	dicom_output_options.done();
	options.done();

	BinaryInputOpenerFromOptions input_opener(
		options,binary_input_options.filename,cin);
	DicomOutputOpenerFromOptions output_opener(
		options,dicom_output_options.filename,cout);

	cerr << binary_input_options.errors();
	cerr << dicom_output_options.errors();
	cerr << options.errors();
	cerr << input_opener.errors();
	cerr << output_opener.errors();

	if (!binary_input_options.good()
	 || !dicom_output_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !output_opener.good()
	 || !options) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< binary_input_options.usage()
			<< dicom_output_options.usage()
			<< " [-image imagenumberfrom0]"
			<< " [-v|-verbose]"
			<< " [" << MMsgDC(InputFile)
				<< "[" << MMsgDC(OutputFile) << "]]"
			<< " <" << MMsgDC(InputFile)
			<< " >" << MMsgDC(OutputFile)
			<< endl;
		exit(1);
	}

	BinaryInputStream bin(*(istream *)input_opener);
	DicomOutputStream dout(*(ostream *)output_opener,
		dicom_output_options.transfersyntaxuid,
		dicom_output_options.usemetaheader,
		dicom_output_options.useimplicitmetaheader,
		dicom_output_options.addtiff);

	ManagedAttributeList list;

	TransferSyntax transfersyntax(
		dicom_output_options.transfersyntaxuid
		? dicom_output_options.transfersyntaxuid
		: DefaultTransferSyntaxUID);

	TextOutputStream log (cerr);
	if (verbose) log << "Converting ..." << endl;

	IMTN_Conversion convertor(bin,cerr);
	convertor.convertAll(&list,&transfersyntax,imagenumber);

	if (!usualManagedAttributeListWrite(list,dout,
		dicom_output_options,log,verbose)) return 1;

	return 0;
}

